6964: [Agent] Automatically open base module for accessing Unsafe on JDK 11+

gunnarmorling opened the pull request:
So here's a fresh take on the `Unsafe` / `add-opens` issue for JMC Agent. This still is using `Unsafe`, but I'm using the instrumentation API for opening up the base module within the agent itself, instead of requiring the user to do so. Ideally, we'd still get away from `Unsafe`, but this make the agent easier to use. This is still a draft, would do some more polishing if you all think it's the right approach. Feedback welcome!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | linux | mac | win |
| --- | ----- | ----- | ----- |
| Build / test | ❌ (1/1 failed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) |

**Failed test task**
- [linux](https://github.com/gunnarmorling/jmc/runs/1470660959)

### Issue
 * [JMC-6964](https://bugs.openjdk.java.net/browse/JMC-6964): [Agent] Automatically open base module for accessing Unsafe on JDK 11+


### Reviewers
 * [Marcus Hirt](https://openjdk.java.net/census#hirt) (@thegreystone - **Reviewer**) ⚠️ Review applies to 5d4792ff4c6efebcf60775264e7874bc185a2ddf


### Download
`$ git fetch https://git.openjdk.java.net/jmc pull/152/head:pull/152`
`$ git checkout pull/152`

gunnarmorling commented:
Note that based on this I think it should also be possible to obtain a `MethodHandles.Lookup` with the private access in target packages which is required for using the publicly supported `MethodHandles` API instead of `Unsafe`.

gunnarmorling commented:
@thegreystone, so that's ready for review from my PoV.

While not removing the need for `Unsafe` on JDK 11+ (there's my other WIP PR for that), it's a good intermediary improvement over the current situation, as it avoids the need for the user to specify the `--add-opens` option themselves. Could you log an issue so I can reference it in the commit message? Thanks!

guruhb commented:
Please change title of this pull request matching to https://bugs.openjdk.java.net/browse/JMC-6964

gunnarmorling commented:
> Please change title of this pull request matching to https://bugs.openjdk.java.net/browse/JMC-6964

Done; thx for logging the issue!

thegreystone commented on agent/src/main/java/org/openjdk/jmc/agent/Agent.java:
```diff
@@ -71,9 +75,44 @@
+						Collections.emptyMap() // extraProvides
+				);
+			} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+				System.out.println(
```
Perhaps moving the openUnsafePackage method out of the Agent class might be nice to keep the agent class clean. Also, since you use the same pattern in the VersionUtils, perhaps some static printAndExit(Exception e, int errorCode, String message) too. And perhaps use different exit codes for failing to open up the package vs not being able to determine the version.

Perhaps you could have an AgentStartupUtils under the agent.impl package and move the VersionUtils methods there too? These are only used when starting the agent.

gunnarmorling replied:
Good ideas, will rework based on that. Re VersionUtils, there's similar demands for that elsewhere; seeing now I probably should merge this with `VersionResolver`.

gunnarmorling commented on agent/src/main/java/org/openjdk/jmc/agent/util/VersionUtils.java:
```diff
@@ -0,0 +1,86 @@
+			}
+
+		} catch (Exception e) {
+			System.out.println("Could not identify Java version. The agent will not work"); //$NON-NLS-1$
```
Tbh. I just took inspiration from `TypeUtils#getUnsafeDefineClassMethod(Object)`, so I thought it's what should happen. What's the general stance for uncoverable exceptions in the agent? Should the app just run normally, ignoring any errors in the agent? If so, a general catch-all handler in the agent's main method might be useful.

gunnarmorling commented:
Thanks for reviewing, @thegreystone. Will update in a bit.

gunnarmorling commented:
@thegreystone, reworked and pushed an update along the lines we discussed:

- moved module opens handling to separate class
- centralized handling of Java version
- don't exiting any longer upon error (this would happen later on, as before, when the module opening failed for some reason)

// CC @Josh-Matsuoka

guruhb commented:
@gunnarmorling can you assign jbs https://bugs.openjdk.java.net/browse/JMC-6964 to your name ?

gunnarmorling commented:
@guruhb, I'm not enrolled with that Jira instance yet. Would you have any pointers on how to register? Thanks!

thegreystone commented:
Gunnar will need to become an author first. I'll sponsor.

gunnarmorling commented:
Ok, sounds great :) Can you just tell me what I need to do for that? Not to bug you, but my knowledge on the process is non-existent. Thanks for any pointers!

thegreystone commented:
To become an author, it's standard OpenJDK rules:

"Becoming an Author
Any Contributor may request the Author role by contacting the relevant Project Lead. References to at least two sponsored contributions to the Project are required. An Author receives the following privileges above those of a Contributor:

An OpenJDK username, which allows write access to the code-review server and (eventually) other infrastructure services, and

The ability to create Mercurial changesets directly, with the Author's OpenJDK username recorded as the author of the Mercurial changeset. This is a more efficient process than the indirect one, where a patch must be sent to an existing Committer and the changeset comment lists the Contributor's e-mail address in the Contributed-by: line.

An Author who is not a Committer still requires the assistance of a Committer to push changesets, but the Author role is nonetheless a step above being an unregistered Contributor.

Project Leads are encouraged to grant the Author role only to Contributors who appear likely to continue to contribute for some time so that they will eventually be nominated to the Committer role."

http://openjdk.java.net/projects/

(I guess the Hg wording needs a bit of an update.)

gunnarmorling commented:
Ok, cool. So should I request that role myself, or do you want to follow that flow described at "Appointing a Contributor to be an Author":

> The Project Lead may appoint any Contributor who has made at least two sponsored contributions to the Project to be a Author by notifying the registrar.

thegreystone commented:
> Project Leads are encouraged to grant the Author role only to Contributors who appear likely to continue to contribute for some time so that they will eventually be nominated to the Committer role.

If you believe this holds true, I'd be happy to appoint you right away! :)

gunnarmorling commented:
While I rather should not make any promises, I'd definitely like to stay involved with JMC Agent; e.g. I'd like to make the converter mechanism work. It's your call of course :) No need to rush from my side really, I don't think I _need_ to have that issue assigned in JBS as suggested by @guruhb in order to move forward with this PR?

gunnarmorling commented:
Hey all, so can I do anything else in order to move this one forward? Thanks for any feedback!

thegreystone commented:
Nope. It's just waiting on me finding some time.  Sorry for the delay.

thegreystone commented on agent/src/main/java/org/openjdk/jmc/agent/util/ModuleUtils.java:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020 Red Hat Inc. All rights reserved.
```
Comma after year.

gunnarmorling replied:
Will add.

gunnarmorling replied:
Comma added.

thegreystone commented on agent/src/main/java/org/openjdk/jmc/agent/util/ModuleUtils.java:
```diff
@@ -0,0 +1,82 @@
+		VersionUtils.getFeatureVersion().ifPresent(featureVersion -> {
+			if (featureVersion >= 11) {
+
+				Method redefineModule = getRedefineModuleMethod();
```
I am not great with the access controller and running with a security manager, but are these reflective checks and module openings something that would benefit from doPrivileged?

Even the (JDK) implementation of the Instrumentation interface is using doPrivileged to do reflective stuff (getting access to premain and agentmain). 
https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java#L416
https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java#L509

gunnarmorling replied:
Is anyone using the security manager still these days ;)

If we wanted to support it, maybe let's tackle that via a separate issue? I think this will need more holistic consideration, there's other places too where a privileged action would be needed (e.g. when obtaining the `Unsafe`). 

Personally, I don't think it's worth the effort and would wait for user feedback/requirement for it. As one data point, we added SM support for Hibernate Validator and never heard anything about it (well, perhaps it just worked). But we didn't do it for Hibernate ORM, and we didn't hear anything there either. Testing for SM is a rather painful experience, too, as you have JUnit and things like that on the callstack and need to accomodate for that.

thegreystone commented:
Looks good! :)

> While I rather should not make any promises, I'd definitely like to stay involved with JMC Agent; e.g. I'd like to make the converter mechanism work. It's your call of course :)

I'm almost done with that (and a big cleanup). :)

gunnarmorling commented:
> I'm almost done with that (and a big cleanup). :)

Oh, converters, you mean? That's great. Less to do for me then :) I got one general thought on that though, will start a thread on the ML.

thegreystone commented:
I did a partial commit here so that you can see where I'm heading with this:
https://github.com/thegreystone/jmc/tree/agentdefaultconverters

thegreystone commented:
(Never mind the "default" part. Had an idea that just introduced too much complexity - ditched it. :) )

thegreystone reviewed (approved):
Ok. Let's wait with the security manager stuff until someone requests it then.

gunnarmorling commented:
Ready to go from my PoV.

/integrate

guruhb commented:
/sponsor
